eth/db/chain_gaps.py

```diff
@@ -12,16 +13,74 @@ class GapChange(enum.Enum):
     NewGap = enum.auto()
     GapFill = enum.auto()
     GapSplit = enum.auto()
-    GapShrink = enum.auto()
+    GapLeftShrink = enum.auto()
+    GapRightShrink = enum.auto()
     TailWrite = enum.auto()
 
 
-GAP_WRITES = (GapChange.GapFill, GapChange.GapSplit, GapChange.GapShrink)
+GAP_WRITES = (
+    GapChange.GapFill,
+    GapChange.GapSplit,
+    GapChange.GapLeftShrink,
+    GapChange.GapRightShrink,
+)
 GENESIS_CHAIN_GAPS = ((), BlockNumber(1))
 
 GapInfo = Tuple[GapChange, ChainGaps]
 
 
+@to_tuple
+def _join_overlapping_gaps(unjoined_gaps: Tuple[BlockRange, ...]) -> Iterable[BlockRange]:
+    """Input must already be sorted"""
+    unyielded_low = None
+    unyielded_high = None
+    for low, high in unjoined_gaps:
+        if unyielded_high is not None:
+            if low < unyielded_low:
+                raise ValidationError(f"Unsorted input! {unjoined_gaps!r}")
+            elif unyielded_low <= low <= unyielded_high + 1:
+                unyielded_high = max(high, unyielded_high)
+                continue
+            else:
+                yield unyielded_low, unyielded_high
+
+        unyielded_low = low
+        unyielded_high = high
+
+    if unyielded_high is not None:
+        yield unyielded_low, unyielded_high
+
+
+def insert_gap(decanonicalized: BlockNumber, base_gaps: ChainGaps) -> ChainGaps:
```

Comment:
This insert gap feature is big enough that it belongs in a separate PR, which will be a bit more reviewable.